### ISSUE TITLE
When a file is changed, update parsing and resolution diagnostics for dependent files

### ISSUE THREAD

**Issue by keyboardDrummer** [dafny-lang/ide-vscode#99]
### Background

Currently the Dafny IDE considers each opened file separately, even when one open file depends on another one. The IDE keeps separate state and runs a separate compilation for each opened file.

Related issue: https://github.com/dafny-lang/ide-vscode/issues/82

### Change

Enable customers to define which `.dfy` files belong together as part of a Dafny 'project'. When any file in a project is changed, update parsing and resolution diagnostics for all files in the project. Verification diagnostics are updated according to the configuration of automatic verification as specified in https://github.com/dafny-lang/ide-vscode/issues/97.

Which files a project contains can be done by specifying a root folder. All .dfy files someone under the folder, nested or not, will be part of the project.

As long as any file in a Dafny project is open, show diagnostics for all files in that project, regardless of whether they're in files that are not opened or not.

The Dafny IDE should watch for file changes that could affect which Dafny projects the currently opened files are in.

### Not in scope

- Allow configuring file filters to select which files within a root folder are part a project.
- Enable specifying the Dafny version a project is built.
- Enable configuring all sorts of project related information such as author, version etc.

### Prerequisites

- Add different modes for automatically triggering verification, and enable manually triggering verification: https://github.com/dafny-lang/ide-vscode/issues/97

### Implementation suggestions

- When the language server starts processing a file, traverse up the file's path to see if a `dfyproject.toml` is present. If such a file is found, this defines a Dafny project which the currently processed file is a part of. 
- When any file in a Dafny project is changed, start a new compilation task that includes all source files in the project, and that cancels any parsing and resolution tasks that have become stale by the change. If there are no caching features present to determine whether a task has become stale, cancel it and recompute.

**Comment by camrein** [comment]
A possible follow-up is the general handling of diagnostics. Currently, diagnostics are hidden when closing the respective document. If we treat the workspace as a single project, I think it makes sense to no longer hide the diagnostics for closed files.

**Comment by emina** [comment]
It would be great to fix this.  In my project, this manifested as fixing a bug in a module A, but having all of the modules that import A still showing the import as underlined.  The only workaround I found was to restart VSCode.

**Comment by camrein** [comment]
> It would be great to fix this. In my project, this manifested as fixing a bug in a module A, but having all of the modules that import A still showing the import as underlined. The only workaround I found was to restart VSCode.

You've probably already tried it. It should be enough to reopen the file only. Or make a little change to the file to trigger re-parsing of the opened file and all its includes.

I had some thoughts regarding this in the past how to improve this experience. I'm sharing these since it might be an idea of how to cope with this in the meantime.
The method `TryParseIncludesOfModule` of [DafnyLangParser.cs](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyLanguageServer/Language/DafnyLangParser.cs) resolves the full list of includes, i.e., direct and indirect includes. Therefore, we know all the dependencies of the Dafny document. If store this information in the `DocumentDatabase`, we've got the ability to track which documents might be affected by a change of a document; thus, can re-trigger the parsing and resolution where necessary.

However, this will have an effect on the responsiveness since we cannot process documents in parallel. To counter this, I suggest introducing debouncing for updates of this kind. For example, if the user didn't apply a change in the past seconds, update the files affected by the last changes.
Another option would be using a (priority) queue. User-introduced changes will always be on the front of the queue. But we'd need to give some more thoughts (or a small prototype) here since I'm not too sure yet how well it plays with the cancellation logic we have.
A third option that came into my mind just as I was typing: Create a dedicated thread for processing updates caused by the change of an included file. Although we cannot process changes in parallel, we could use/create synchronization primitives to ensure that user changes will always be processed first. I think this has the least impact on the current code-base.

**Comment by keyboardDrummer** [comment]
Closing in favor of https://github.com/dafny-lang/dafny/issues/2328
